experimental/JuLie/partitions.jl

```diff
@@ -0,0 +1,727 @@
+```julia-repl
+julia> @btime partitions(Int8(90));
+3.376 s (56634200 allocations: 6.24 GiB)
+
+julia> @btime ascending_partitions(Int8(90),alg="ks");
+3.395 s (56634200 allocations: 6.24 GiB)
+
+julia> @btime ascending_partitions(Int8(90),alg="m");
+3.451 s (56634200 allocations: 6.24 GiB)
+```
+"""
+function ascending_partitions(n::IntegerUnion; alg="ks")
+
+  #Argument checking
+  n >= 0 || throw(ArgumentError("n >= 0 required"))
+
+  # Use type of n
+  T = typeof(n)
+
+  # Some trivial cases
+  if n==0
+    return Vector{T}[ [] ]
+  elseif n==1
+    return Vector{T}[ [1] ]
+  end
+
+  # Now, the algorithm starts
+  if alg=="ks"
+    P = Vector{T}[]    #this will be the array of all partitions
+    a = zeros(T, n)
+    k = 2
+    y = n-1
+    while k != 1
+      k -= 1
+      x = a[k] + 1
+      while 2*x <= y
+        a[k] = x
+        y -= x
+        k += 1
+      end
+      l = k + 1
+      while x <= y
+        a[k] = x
+        a[l] = y
+        push!(P, a[1:l])
+        x += 1
+        y -= 1
+      end
+      y += x - 1
+      a[k] = y + 1
+      push!(P, a[1:k])
+    end
+    return P
+
+  elseif alg=="m"
+    P = Vector{T}[]    #this will be the array of all partitions
+    a = zeros(T, n)
+    k = 1
+    x = 1
+    y = n-1
+    while true
+      while 3*x <= y
+        a[k] = x
+        y = y-x
+        k += 1
+      end
+      t = k + 1
+      u = k + 2
+      while 2*x <= y
+        a[k] = x
+        a[t] = x
+        a[u] = y - x
+        push!(P, a[1:u])
+        p = x + 1
+        q = y - p
+        while p <= q
+          a[t] = p
+          a[u] = q
+          push!(P, a[1:u])
+          p += 1
+          q -= 1
+        end
+        a[t] = y
+        push!(P, a[1:t])
+        x += 1
+        y -= 1
+      end
+      while x<=y
+        a[k] = x
+        a[t] = y
+        push!(P, a[1:t])
+        x += 1
+        y -= 1
+      end
+      y += x-1
+      a[k] = y+1
+      push!(P, a[1:k])
+      k -= 1
+
+      # I think there's a mistake in the publication
+      # because here k could be zero and then we access
+      # a[k].
+      # That's why I do a while true and check k > 0 here.
+      if k == 0
+        break
+      else
+        x = a[k] + 1
+      end
+    end
+    return P
+  else
+    error("alg must be either ks or m")
+  end
+
+end
+
+
+
+@Markdown.doc """
+    partitions(m::IntegerUnion, n::IntegerUnion, l1::IntegerUnion, l2::IntegerUnion; z=0)
+
+A list of all partitions of an integer ``m ≥ 0`` into ``n ≥ 0`` parts with
+lower bound ``l1 ≥ 0`` and upper bound ``l2 ≥ l1`` for the parts. There are
+two choices for the parameter z:
+* z=0: no further restriction (*default*);
+* z=1: only distinct parts.
+The partitions are produced in *decreasing* order.
+
+The algorithm used is "parta" in [RJ76](@cite), de-gotoed from old ALGOL code by E. Thiel!
+"""
+function partitions(m::IntegerUnion, n::IntegerUnion, l1::IntegerUnion, l2::IntegerUnion; z::IntegerUnion=0)
```

Comment:
If `z` only allows for two choices, there is no point in allowing it to be `IntegerUnion`, it can go back to being `Int`.

However, why isn't this a `Bool` with a speaking name? E.g.
```suggestion
function partitions(m::IntegerUnion, n::IntegerUnion, l1::IntegerUnion, l2::IntegerUnion; only_distinct_parts::Bool = false)
```